### PR TITLE
[bugfix]Fix bug in Lora checkpoint saving step

### PR DESCRIPTION
**Pull request by leisuzz**
# What does this PR do?

Fix bug in checkpoint saving steps:
1. Free variable referenced before assignment in the original code if it is only `elif isinstance(model, type(unwrap_model(text_encoder)))`
2. IndexError: pop from empty list

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by leisuzz** [comment]
@sayakpaul Please take a look at this PR, I'm having trouble in dreambooth lora checkpoint saving. After these implementation, the issue has been solved

**Comment by sayakpaul** [comment]
Just so I understand, this is applicable only when we're using DeepSpeed?

**Comment by leisuzz** [comment]
@sayakpaul I'm using deepspeed to train the model, I'm not sure whether it works or not without deepspeed

**Comment by sayakpaul** [comment]
Same comments from your other PR that you closed apply here. We're missing:

* Incorporating LoRA/model loading support when using DeepSpeed. 
* Other necessary checks when using DeepSpeed. 

As mentioned in the other PR of yours, if you go [here](https://github.com/a-r-r-o-w/cogvideox-factory/blob/main/training/cogvideox_image_to_video_lora.py#L561) and search for `DistributedType.DEEPSPEED`, the outline of the changes should become evident and they will help you understand what we're missing in this PR. Without those changes, it'd be hard to merge this PR.

**Comment by leisuzz** [comment]
@sayakpaul I don' have any issue related with loading with deepspeed, the issue only occurs in saving steps

**Comment by sayakpaul** [comment]
We have extensively tested the changes I mentioned in my comment. Hence, I kindly ask you to propagate them.

**Comment by leisuzz** [comment]
@sayakpaul I've updated the save mode for the correct version, please take a look. This should work properly for loading function as well

File: examples/dreambooth/train_dreambooth_lora_flux.py [hunk 1]
```diff
@@ -1785,7 +1788,7 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
                 progress_bar.update(1)
                 global_step += 1
 
-                if accelerator.is_main_process:
+                if accelerator.is_main_process or accelerator.distributed_type == DistributedType.DEEPSPEED:
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux.py:1791]:
```suggestion
                if accelerator.distributed_type == DistributedType.DEEPSPEED or accelerator.is_main_process:
```

**Review by sayakpaul — COMMENTED** [review]
Let's keep the changes limited to the Flux, script perhaps?

**Comment by leisuzz** [comment]
@sayakpaul The issue happened both in sd3 and flux actually, I've also fixed the issue with distributed in condition with checkpoints_total_limit is not none

**Comment by leisuzz** [comment]
@sayakpaul I've fixed both loading and saving step, please take a look

**Comment by sayakpaul** [comment]
We'd prefer to tackle this one script at a time. So, please keep the changes to a single script, only.

**Comment by leisuzz** [comment]
@sayakpaul I've edited and now the changes are only based on train_dreambooth_lora_sd3.py (sd3 and flux are similar and I tested both of them), please take a look. Let me know if that's ok for me to edit on flux as well.

File: examples/dreambooth/train_dreambooth_lora_sd3.py [hunk 2]
```diff
@@ -1315,54 +1322,55 @@ def save_model_hook(models, weights, output_dir):
...
+                    text_encoder_two_ = model
+                else:
+                    raise ValueError(f"unexpected save model: {model.__class__}")
 
-        # Make sure the trainable params are in float32. This is again needed since the base models
-        # are in `weight_dtype`. More details:
-        # https://github.com/huggingface/diffusers/pull/6514#discussion_r1449796804
-        if args.mixed_precision == "fp16":
-            models = [transformer_]
+            lora_state_dict = StableDiffusion3Pipeline.lora_state_dict(input_dir)
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sd3.py:1342]: We don't have to do LoRA here.

File: examples/dreambooth/train_dreambooth_lora_sd3.py [hunk 3]
```diff
@@ -1315,54 +1322,55 @@ def save_model_hook(models, weights, output_dir):
...
+                    )
             if args.train_text_encoder:
-                models.extend([text_encoder_one_, text_encoder_two_])
-            # only upcast trainable parameters (LoRA) into fp32
-            cast_training_params(models)
+                # Do we need to call `scale_lora_layers()` here?
+                _set_state_dict_into_text_encoder(lora_state_dict, prefix="text_encoder.", text_encoder=text_encoder_one_)
+
+                _set_state_dict_into_text_encoder(
+                    lora_state_dict, prefix="text_encoder_2.", text_encoder=text_encoder_two_
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sd3.py:1362]: We don't have to do LoRA here.

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: examples/dreambooth/train_dreambooth_lora_sd3.py (hunk 3 above)
Comment (leisuzz) [reply examples/dreambooth/train_dreambooth_lora_sd3.py:1362]: I only wrapped the code under if not accelerator.distributed_type == DistributedType.DEEPSPEED:

Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sd3.py:1362]: Yeah but it's not needed.

Comment (leisuzz) [reply examples/dreambooth/train_dreambooth_lora_sd3.py]: I've edited it, please take a look

**Review by sayakpaul — COMMENTED** [review]
Just a few more changes.

File: examples/dreambooth/train_dreambooth_lora_sd3.py [hunk 4]
```diff
@@ -1315,54 +1322,32 @@ def save_model_hook(models, weights, output_dir):
...
-            cast_training_params(models)
+            # Make sure the trainable params are in float32. This is again needed since the base models
+            # are in `weight_dtype`. More details:
+            # https://github.com/huggingface/diffusers/pull/6514#discussion_r1449796804
+            if args.mixed_precision == "fp16":
+                models = [transformer_]
+                if args.train_text_encoder:
+                    models.extend([text_encoder_one_, text_encoder_two_])
+                # only upcast trainable parameters (LoRA) into fp32
+                cast_training_params(models)
```
[the remark this answers is not in the document]
Comment (leisuzz) [reply examples/dreambooth/train_dreambooth_lora_sd3.py]: Just tested them, both work

**Comment by leisuzz** [comment]
@sayakpaul Please take a look at this newest version, thanks

**Comment by sayakpaul** [comment]
@leisuzz where you not using a non LoRA version of the script? I think I got confused because of the repeated reviews. 

Maybe we could close this PR and I could instead open a PR with you as a co-author?

**Comment by leisuzz** [comment]
@sayakpaul Sorry about this, I will close this PR then

**Comment by sayakpaul** [comment]
@leisuzz thanks for your understanding. Could you please provide your GitHub commit email address so that I can add you as a [co-author](https://docs.github.com/articles/creating-a-commit-with-multiple-authors)?

**Comment by leisuzz** [comment]
My email address is: jiangshuonb@gmail.com

**Comment by leisuzz** [comment]
@sayakpaul My email address is: jiangshuonb@gmail.com. Thanks for your help!

**Pull request closed without merging**